### PR TITLE
Deps save and drop dependencies

### PR DESCRIPTION
**Pull request by gabrielwol**
update public.deps_save_and_drop_dependencies to include "ALTER OWNER" and "ADD (UNIQUE) CONSTRAINT" statements in deps_saved_ddl table. 
Also fluffed and added to version control!

### CODE REVIEW

**Comment by gabrielwol** [comment]
@chmnata I think we'll need to use this function for current [miovision re-partitioning ](https://github.com/CityofToronto/bdit_data-sources/pull/735) so your review would be very helpful! 
I tried to organize it so the original function + fluff is in first two commits, so if you just look at the remaining 5 commits it's easier to see just the substantial differences:
https://github.com/CityofToronto/bdit_pgutils/pull/9/files/709cff5ce69a7537cca3efd1abf4235881093f31..2630845fe0415afb19c8dd055be169dcf9b30c3d

**Comment by chmnata** [comment]
Looks good! Have you tested the function yet? I think it would be good to add a comment for the function to describe what they do, including the statements it can handle, e.g. alter owners, grant permissions, save table and column comments, index and ......

**Comment by gabrielwol** [comment]
@chmnata I added a looong comment. 
I also added a dryrun option that Raph was interested in, which allows a user to run the function without actually dropping deps. 

I tested in my personal schema (see `gwolofs.deps_save_and_drop_dependencies`) and based on that experience also reorganized the inserts so that the CREATE statement comes first, before any modifications (ALTER/GRANT/COMMENT).

**Comment by chmnata** [comment]
amazing comment!!! I will test it out with a dry run with a couple of tables, if it looks good we can mergeee :meow_salute:

File: create-function-deps_restore_dependencies.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
+CREATE OR REPLACE FUNCTION public.deps_restore_dependencies(p_view_schema IN VARCHAR, p_view_name IN VARCHAR)
+RETURNS NULL
```
Comment (chmnata) [inline create-function-deps_restore_dependencies.sql:2]: is this supposed to be `void` ? Its not recognizing NULL as an optio. Or is it supposed to be `RETURNS NULL ON NULL INPUT` ?

File: create-function-deps_save_and_drop_dependencies.sql [hunk 2]
```diff
@@ -0,0 +1,212 @@
+CREATE OR REPLACE FUNCTION public.deps_save_and_drop_dependencies(p_view_schema IN VARCHAR, p_view_name IN VARCHAR, dryrun BOOLEAN default False)
```
Comment (chmnata) [inline create-function-deps_save_and_drop_dependencies.sql:1]: this one looks good!

File: create-function-deps_save_and_drop_dependencies.sql [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        WHEN v_curr.obj_type = 'm' THEN 'MATERIALIZED VIEW'
+    END || ' ' || v_curr.obj_schema || '.' || v_curr.obj_name || ';';
+END IF;
+
+END loop;
+
+END;
+$$;
+
+ALTER FUNCTION public.deps_save_and_drop_dependencies OWNER TO natalie;
```
Comment (chmnata) [inline create-function-deps_save_and_drop_dependencies.sql:191]: Can you add (VARCHAR, VARCHAR, BOOLEAN) to the end of the function as well? To make it explicit

File: create-function-deps_restore_dependencies.sql (hunk 1 above)
Comment (gabrielwol) [reply create-function-deps_restore_dependencies.sql:2]: right, void. I had to change that on the other one during testing.

**Comment by gabrielwol** [comment]
In my testing I found the permissions aren't saved for mat views.
some solutions offered here: https://stackoverflow.com/questions/38728548/list-grants-and-privileges-for-a-materialized-view-in-postgresql

**Comment by chmnata** [comment]
Just documenting our IRL discussion and next steps:
- Separate dry run to another function
- Delete existing rows related to the table we are running the function on in the deps table before inserting again

Out of scope in this issue:
- Include create trigger rules

**Comment by gabrielwol** [comment]
I changed the main function into deps_save_and_drop_dependencies **_dryrun** with EXECUTE to bdit_humans and made a new function into `deps_save_and_drop_dependencies` with default dryrun = False with permissions only to bdit_admins. I think this is good enough because if a human tried to run _dryrun with dryrun = False they would run into permission issues on the drop statements. 
I also added a delete statement to clear `deps_saved_ddl` at the beginning.

File: create-function-deps_save_and_drop_dependencies_dryrun.sql [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        WHEN v_curr.obj_type = 'm' THEN 'MATERIALIZED VIEW'
+    END || ' ' || v_curr.obj_schema || '.' || v_curr.obj_name || ';';
+END IF;
+
+END loop;
+
+END;
+$$;
+
+ALTER FUNCTION public.deps_save_and_drop_dependencies_dryrun(VARCHAR, VARCHAR, BOOLEAN) OWNER TO bdit_admins;
```
Comment (chmnata) [inline create-function-deps_save_and_drop_dependencies_dryrun.sql:205]: Can you change to `dbadmin` instead?

File: create-function-deps_save_and_drop_dependencies.sql [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    COST 100
+AS $$ 
+SELECT public.deps_save_and_drop_dependencies_dry_run(
+    p_view_schema,
+    p_view_name,
+    False --dryrun = False
+);
+$$;
+
+GRANT EXECUTE ON FUNCTION public.deps_save_and_drop_dependencies(VARCHAR, VARCHAR) TO bdit_admins;
```
Comment (chmnata) [inline create-function-deps_save_and_drop_dependencies.sql:15]: Can you change to `dbadmin` instead?

File: create-function-deps_restore_dependencies.sql [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+DELETE FROM public.deps_saved_ddl
+WHERE
+    deps_view_schema = p_view_schema
+    AND deps_view_name = p_view_name;
+
+END;
+$$;
+
+ALTER FUNCTION public.deps_restore_dependencies(VARCHAR, VARCHAR) OWNER TO natalie;
```
Comment (chmnata) [inline create-function-deps_restore_dependencies.sql:31]: should this still be me ?

**Comment by chmnata** [comment]
I think everything looks good! Do you mind sending me the block of code you use for testing? Or maybe you can show it to me tmr with screenshare.

File: create-function-deps_restore_dependencies.sql (hunk 6 above)
Comment (gabrielwol) [reply create-function-deps_restore_dependencies.sql:31]: changed to dbadmin

**Comment by gabrielwol** [comment]
FYI Here is the code block I used to test: 

```sql
--DROP MATERIALIZED VIEW gwolofs.employees_test_mat_view;
CREATE MATERIALIZED VIEW gwolofs.employees_test_mat_view AS SELECT * FROM gwolofs.employees;
CREATE UNIQUE INDEX ON gwolofs.employees_test_mat_view USING btree(employee_id);
GRANT ALL ON gwolofs.employees_test_mat_view TO natalie;
COMMENT ON MATERIALIZED VIEW gwolofs.employees_test_mat_view IS 'Test';
COMMENT ON COLUMN gwolofs.employees_test_mat_view.department IS 'Test';

--dry run example: 
SELECT gwolofs.deps_save_and_drop_dependencies_dryrun('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");
SELECT deps_ddl_to_run FROM gwolofs.deps_saved_ddl WHERE deps_view_schema = 'gwolofs' AND deps_view_name = 'employees' ORDER BY deps_id;

--not dry run: 
SELECT gwolofs.deps_save_and_drop_dependencies('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");
SELECT gwolofs.deps_restore_dependencies('gwolofs'::text COLLATE pg_catalog."C", 'employees'::text COLLATE pg_catalog."C");

/*triggers on views is not supported: 
CREATE OR REPLACE TRIGGER transform_trigger
    INSTEAD OF INSERT
    ON gwolofs.employees_test_mat_view
    FOR EACH ROW
    EXECUTE FUNCTION here.here_insert_trigger();
*/
```

**Comment by chmnata** [comment]
Amazing, thanks! Merging

**Pull request merged**